### PR TITLE
docker: Make the 'Change resource limits' dialog actually work

### PR DESCRIPTION
**Pull request by stefwalter**
Although this only affects a running container as you would sorta
expect ... it now actually works.

We fix how errors are displayed in the dialog and progress is
displayed.

Also fix races between changing the three settings.

We have to be careful in which order we set the memory and memory swap
limits. In particular when setting -1 we have to clear the swap
limit first, and then go for the memory limit.

### CODE REVIEW

**Comment by stefwalter** [comment]
Needs an integration test. And likely a way to type the actual number, the same way as the storage dialog.

**Comment by dperpeet** [comment]
issue for this is #3439

**Comment by stefwalter** [comment]
Added text fields and tests.

**Comment by stefwalter** [comment]
Marking priority so we can get this into the next release.

File: pkg/docker/client.js [hunk 1]
```diff
@@ -573,20 +573,26 @@ define([
...
         }
 
         this.change_memory_limit = function change_memory_limit(id, value) {
+            var cgroup = this.containers[id].CGroup;
             if (value === undefined || value <= 0)
                 value = -1;
-            return change_cgroup("memory", this.containers[id].CGroup, "memory.limit_in_bytes", value);
+
+            /* The order in which we set memory.memsw and memory is important. */
+            if (value == -1) {
```
Comment (dperpeet) [inline pkg/docker/client.js]: should this use `===`?

File: pkg/docker/details.js [hunk 2]
```diff
@@ -68,12 +69,9 @@ define([
                     self.cpu_priority.value = info.CpuPriority || undefined;
                 }).
                 find(".btn-primary").on("click", function() {
-                    self.client.change_memory_limit(self.container_id, self.memory_limit.value);
-                    var swap = self.memory_limit.value;
-                    if (!isNaN(swap))
-                        swap *= 2;
-                    self.client.change_swap_limit(self.container_id, swap);
```
Comment (mvollmer) [inline pkg/docker/details.js]: The `change_swap_limit` function can be removed as well.

**Comment by stefwalter** [comment]
@mvollmer I think reporting data accurately on fedora-24 should be a different issue and pull request.

File: pkg/docker/client.js (hunk 1 above)
Comment (stefwalter) [inline pkg/docker/client.js]: Done.

File: pkg/docker/details.js (hunk 2 above)
Comment (stefwalter) [inline pkg/docker/details.js]: Removed

**Comment by mvollmer** [comment]
> @mvollmer I think reporting data accurately on fedora-24 should be a different issue and pull request.

Totally.

**Comment by mvollmer** [comment]
Looks good.

**Comment by stefwalter** [comment]
> On rhel-7:  /test-setup terminated with SIGSEGV FAIL: test-setup 9 /setup/commit/passwd1

That's another bug that needs work to debug. I'll try to get around to it, but it doesn't have to do with this pull request.

**Comment by stefwalter** [comment]
Trying to debug the Memoryswap issue on debian.

**Comment by stefwalter** [comment]
On debian memory-swap is not supported: 

```
root@unassigned-hostname:~# docker run -ti --memory=5M --memory-swap=10M busybox /bin/sh
Error response from daemon: You should always set the Memory limit when using Memoryswap limit, see usage.
```

**Comment by stefwalter** [comment]
This information is available via docker /info request.

**Comment by stefwalter** [comment]
Did a bunch more work to disable stuff according to the data in /info ... but ran into another problem.

On debian the cgroups look like this:

```
system.slice/docker/792f2c28d75a843d7db0dea6d59ea4c260ae2ebdd99c429379e25fd0c8035957
```

On Fedora 24 they look like:

```
init.scope/system.slice/docker-792f2c28d75a843d7db0dea6d59ea4c260ae2ebdd99c429379e25fd0c8035957.scope
```

And on Fedora 23 and RHEL they look like:

```
system.slice/docker-792f2c28d75a843d7db0dea6d59ea4c260ae2ebdd99c429379e25fd0c8035957.scope
```

This causes the new tests to fail.

**Comment by stefwalter** [comment]
Added fixes for the various layouts for cgroups on different operating systems.

**Comment by dperpeet** [comment]
tests failing
```
# testResources (check_docker.TestDocker)
#
Wrote TestDocker-testResources-FAIL.png
Journal extracted to TestDocker-testResources-10.111.119.229-FAIL.log
not ok 61 testResources (check_docker.TestDocker)
Traceback (most recent call last):
  File "./verify/check-docker", line 383, in testResources
    b.wait_in_text("#container-details-cpu-row", "512 shares")
```

**Comment by stefwalter** [comment]
I've split out the tests into #4611

**Comment by mvollmer** [comment]
@stefwalter, the fixups for the review comments went to the wrong commit, I think.

**Comment by mvollmer** [comment]
> I've split out the tests into #4611

They are still here, no?

**Comment by stefwalter** [comment]
Removed the last commit, and moved review fixups into the right commit. Tests are in #4611

**Pull request closed without merging**